source/common/config/discovery_grpc_stream.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <functional>
+
+#include "envoy/grpc/async_client.h"
+
+#include "common/common/backoff_strategy.h"
+#include "common/common/token_bucket_impl.h"
+#include "common/config/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+// Oversees communication for gRPC-using xDS: reestablishes the gRPC channel when necessary, and
+// provides rate limiting of requests.
+template <class RequestProto, class ResponseProto>
+class DiscoveryGrpcStream : public Grpc::TypedAsyncStreamCallbacks<ResponseProto>,
+                            Logger::Loggable<Logger::Id::upstream> {
+public:
+  DiscoveryGrpcStream(Grpc::AsyncClientPtr async_client,
```

Comment:
WDYT of making these std::function args instead be pure virtual methods and you have to provide an override for those methods on instantiation rather than supplying them as lambdas? It looks like that might not be too bad at the call-site.

Reply:
Done. Yeah, it turned out to be cleaner this way (assuming that decreasing use of multiple inheritance is not itself a goal).